### PR TITLE
Re-enable group constraints

### PR DESCRIPTION
**Pull request by pschorf**
This change has two commits:
1. Reverts the commit removing fenzo group constraints
2. Adds explicit caching for each group's running cotasks (for rebalancing) and running cotask ids (for matching)

Both of these cases don't necessarily need a `core.cache` cache since they have a fixed number of elements, but I think it's more appropriate as an API than a map or something else.

### CODE REVIEW

File: scheduler/src/cook/mesos/constraints.clj [hunk 1]
```diff
@@ -186,11 +187,17 @@
...
-  (-> (group/group->running-task-set db group)
-      (clojure.set/difference (set (:job/instance job)))))
+  (let [group-uuid (:group/uuid group)
+        db-cotasks (cache/lookup (swap! cotask-cache
+                                        (fn [c]
+                                          (if (cache/has? c group-uuid)
+                                            (cache/hit c group-uuid)
+                                            (cache/miss c group-uuid (group/group->running-task-set db group)))))
+                                 group-uuid)]
+    (clojure.set/difference db-cotasks (set (:job/instance job)))))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/constraints.clj]: Nitpick: can we use `set` after importing `clojure.set`

File: scheduler/src/cook/mesos/constraints.clj [hunk 2]
```diff
@@ -186,11 +187,17 @@
 ;; Group host placement constraints
 
 (defn get-cotasks
-  [db group job]
+  [db group job cotask-cache]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/constraints.clj]: Nitpick: I prefer `group-uuid->running-tasks-cache` as it tells the reader and caller about the mapping of the cache. Also, the cache contains the _running_ cotasks.

File: scheduler/src/cook/mesos/constraints.clj [hunk 3]
```diff
@@ -186,11 +187,17 @@
...
   "Given a group and a job, finds the set of all the running instances that belong to the jobs in
    'group', but do not belong to 'job'"
-  (-> (group/group->running-task-set db group)
-      (clojure.set/difference (set (:job/instance job)))))
+  (let [group-uuid (:group/uuid group)
+        db-cotasks (cache/lookup (swap! cotask-cache
+                                        (fn [c]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/constraints.clj]: This function can be extracted and potentially reused (after the refactor to `running-task-ids-cache` suggested below).

File: scheduler/src/cook/mesos/constraints.clj [hunk 4]
```diff
@@ -211,12 +218,18 @@
    sequence of Fenzo TaskTracker.ActiveTasks, which correspond to ALL the running tasks in group.
    This includes both running tasks and tasks that are being assigned by Fenzo
    in the current cycle."
-  [db group task-id same-cycle-task-ids task-tracker-state]
-  (let [cotask-ids (disj
-                     (clojure.set/union
-                       (group/group->running-task-id-set db group) ; Tasks that are in the database (running or scheduled)
-                       same-cycle-task-ids) ; Tasks in same cycle
-                     task-id)] ; Do not include this task-id
+  [db group task-id same-cycle-task-ids task-tracker-state running-task-ids-cache]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/constraints.clj]: Can we not reuse the same cache (i.e. the `cotask-cache` from `get-cotasks`) and use a `map` to extract the task ids?

File: scheduler/src/cook/mesos/constraints.clj [hunk 5]
```diff
@@ -211,12 +218,18 @@
...
+                                             (if (cache/has? c group-id)
+                                               (cache/hit c group-id)
+                                               (cache/miss c group-id (group/group->running-task-id-set db group)))))
+                                    group-id)
+        cotask-ids (disj (clojure.set/union
+                          db-cotask-ids ; Tasks that are in the database (running or scheduled)
+                          same-cycle-task-ids) ; Tasks in same cycle
+                         task-id)] ; Do not include this task-id
     ; Return a Fenzo TaskTracker.ActiveTask
     (get-cotasks-from-tracker-state task-tracker-state cotask-ids)))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/constraints.clj]: On my inital read this looked like a faulty recursive call to me. The function name is `get-cotasks-from-tracker-state`, but the arguments are different. Can we rename one of these functions?

File: scheduler/src/cook/mesos/constraints.clj [hunk 6]
```diff
@@ -211,12 +218,18 @@
...
-                       (group/group->running-task-id-set db group) ; Tasks that are in the database (running or scheduled)
-                       same-cycle-task-ids) ; Tasks in same cycle
-                     task-id)] ; Do not include this task-id
+  [db group task-id same-cycle-task-ids task-tracker-state running-task-ids-cache]
+  (let [group-id (:group/uuid group)
+        db-cotask-ids (cache/lookup (swap! running-task-ids-cache
+                                           (fn [c]
+                                             (if (cache/has? c group-id)
+                                               (cache/hit c group-id)
+                                               (cache/miss c group-id (group/group->running-task-id-set db group)))))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/constraints.clj]: Nitpick: Can we use just one of `group/group->running-task-set` and `group/group->running-task-id-set`?

File: scheduler/test/cook/test/mesos/rebalancer.clj [hunk 7]
```diff
@@ -24,7 +25,8 @@
             [cook.mesos.util :as util]
             [cook.test.testutil :refer (restore-fresh-database! create-dummy-job create-dummy-instance
                                                                 create-dummy-group init-offer-cache)]
-            [datomic.api :as d :refer (q)]))
+            [datomic.api :as d :refer (q)]
+            [clojure.core.cache :as cache]))
```
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/rebalancer.clj]: Should be deleted, `clojure.core.cache` is already imported above.

File: scheduler/test/cook/test/mesos/rebalancer.clj [hunk 8]
```diff
@@ -208,6 +210,7 @@
   (testing "test without group constraints"
     (let [datomic-uri "datomic:mem://test-compute-preemption-decision"
           conn (restore-fresh-database! datomic-uri)
+          cotask-cache (atom (cache/lru-cache-factory {} :threshold 100))
```
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/rebalancer.clj]: Can we please stick to using the `fifo-cache` in the tests too?

File: scheduler/test/cook/test/mesos/rebalancer.clj [hunk 9]
```diff
@@ -438,7 +454,8 @@
                        "sticks" {"HOSTNAME" "sticks"}
                        "bricks" {"HOSTNAME" "bricks"}
                        "rebar" {"HOSTNAME" "rebar"}}
-          offer-cache (init-offer-cache)]
+          offer-cache (init-offer-cache)
+          cotask-cache (atom (cache/lru-cache-factory {} :threshold 100))]
```
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/rebalancer.clj]: Can we please stick to using the `fifo-cache` in the tests too?

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 10]
```diff
@@ -815,6 +816,19 @@
                        vals
                        (reduce #(+ %1 (count (.getTasksAssigned %2))) 0))))))))
 
+(deftest stress-test-constraint
```
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Is this test supposed to assert something?

File: integration/tests/cook/test_basic.py [hunk 11]
```diff
@@ -1335,6 +1335,187 @@ def test_unscheduled_jobs_partial(self):
...
+                unique_reason = [r for r in unscheduled_jobs['reasons'] if r['reason'] ==
+                                 reasons.COULD_NOT_PLACE_JOB][0]
+                self.assertEqual("unique-host-placement-group-constraint",
+                                 unique_reason['data']['reasons'][0]['reason'],
+                                 unique_reason)
+        finally:
+            util.kill_jobs(self.cook_url, uuids)
+
+    @attr('explicit')
+    def test_balanced_host_constraint_cannot_place(self):
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Should we try and make this test non-explicit?

File: scheduler/src/cook/mesos/constraints.clj (hunk 5 above)
Comment (shamsimam) [reply scheduler/src/cook/mesos/constraints.clj]: Oops, my bad. The code collapsed enough to hide the declaration of `get-fenzo-cotasks`

**Comment by pschorf** [comment]
@dposada I think this is ready to merge

File: scheduler/src/cook/mesos/constraints.clj [hunk 12]
```diff
@@ -185,12 +186,23 @@
 
 ;; Group host placement constraints
 
+(defn- get-running-cotasks-from-cache
```
Comment (dposada) [inline scheduler/src/cook/mesos/constraints.clj]: Docstring?

Comment (pschorf) [reply scheduler/src/cook/mesos/constraints.clj]: Added

**Comment by pschorf** [comment]
The patches (aside from the revert) shouldn't have any functional change, just improve the performance.

**Comment by dposada** [comment]
Are we waiting for @wyegelwel to approve before merging?

**Comment by pschorf** [comment]
I think we're ok to go as-is, once the build finishes.

**Pull request merged**